src/hf.f90

```diff
@@ -0,0 +1,318 @@
+         ! it's ao_ort that gets passed into BLAS for diagonalisation
+         ovlp%ao_ort = ovlp%ao
+         call diagonalise(ovlp)
+         ortmat(:,:) = 0.0_p
+         ! ovlp%W holds the eigenvalues s_i of S
+         associate(diag=>ovlp%W)
+            do i = 1, size(ortmat, dim=1)
+               ortmat(i,i) = 1/sqrt(diag(i))
+            end do
+         end associate
+         ! S^-1/2 = U s^{-1/2} U^{\dagger}, the coefficient matrix U is held in ovlp%A
+         ortmat = matmul(ovlp%A, matmul(ortmat, transpose(ovlp%A)))
+         call deallocate_mat_t(ovlp)
+
+         ! ###############################
+         ! Various initialisation routines
+         ! ###############################
+         ! Initial guess Fock matrix is H_core (equivalent to guessing the null matrix for density matrix)
+         ! (Szabo and Ostlund p. 148)
+         call init_mat_t(int_store%core_hamil, fockmat)
+         call init_diis_t(sys, diis)
+
+         ! ####################
+         ! Iterative SCF solver
+         ! ####################
+
+         do iter = 1, sys%scf_maxiter
+            ! F' is the fock matrix in orthogonal AO basis, where F is in the original non-orthogonal basis
+            ! F' = X^T F X
+            fockmat%ao_ort = matmul(transpose(ortmat), matmul(fockmat%ao, ortmat))
+
+            ! F'C' = C'e
+            call diagonalise(fockmat)
+
+            ! C = XC'
+            fockmat%A = transpose(matmul(ortmat, fockmat%A))
+
+            ! D_{\mu\nu} = \sum_{i}^{nocc} C_\mu^i C_\nu^i
+            call build_density(fockmat%A, density, sys%nel/2)
+
+            ! E_el = \sum_{\mu\nu} D_{\mu\nu} (H^{core}_{\mu\nu} + F_{\mu\nu})
+            call update_scf_energy(sys, density, fockmat%ao, st, int_store, conv)
+            if (conv) then
+               write(iunit, '(1X, A)') 'Convergence reached within tolerance.'
+               write(iunit, '(1X, A, 1X, F15.8)') 'Final SCF Energy (Hartree):', st%energy
+               write(iunit, '(1X, A)') 'Orbital energies (Hartree):'
+               do i = size(fockmat%W), 1, -1
+                  write(iunit, '(1X, I3, 1X, F15.8)') i, fockmat%W(i)
+               end do
+
+               ! Copied for return / use in MP2
+               sys%e_hf = st%energy
+               allocate(sys%canon_coeff, source=fockmat%A)
+               allocate(sys%canon_levels, source=fockmat%W)
+               exit
+            end if
+            write(iunit, '(1X, A, 1X, I3, F15.8)') 'Iteration', iter, st%energy
+
+            call build_fock(sys, density, fockmat%ao, int_store)
+
+            ! DIIS update
+            call update_diis(diis, fockmat%ao, density, int_store%ovlp)
+         end do
+
+         if (.not. conv) then
+            write(iunit, '(1X, A)') 'Convergence not reached, please increase maxiter.'
+         end if
+
+         call deallocate_diis_t(diis)
+         deallocate(st%density_old)

```

Comment:
Doing this here seems a bit out of place.  Perhaps have a routine in system.F which does the allocation and deallocation of the bits of state_t?